src/go/rpk/pkg/config/params.go

```diff
@@ -552,3 +568,133 @@ func (c *Config) addUnsetDefaults() {
 		r.AdminAPI.Addresses = []string{"127.0.0.1:9644"}
 	}
 }
+
+// Set allow to set a single configuration property by passing a key value pair
+//
+//   Key:    string containing the yaml property tag, e.g: 'rpk.admin_api'.
+//   Value:  string representation of the value, either single value or partial
+//           representation if a format is passed (yaml / json).
+//   Format: either single, json, or yaml (default: single).
+func (c *Config) Set(key, value, format string) error {
+	if key == "" {
+		return fmt.Errorf("key field must not be empty")
+	}
+	props := strings.Split(key, ".")
+	rv := reflect.ValueOf(c).Elem()
+	found, other, err := getField(props, rv)
+	if err != nil {
+		return err
+	}
+	isOther := other != reflect.Value{}
+
+	field := found
+	if isOther {
+		field = other
+	}
+
+	if field.CanAddr() {
+		i := field.Addr().Interface()
+		in := value
+		switch strings.ToLower(format) {
+		case "yaml", "single", "":
+			if isOther {
+				p := props[len(props)-1]
+				in = fmt.Sprintf("%s: %s", p, value)
+			}
+			err = yaml.Unmarshal([]byte(in), i)
+			if err != nil {
+				return err
+			}
+			return nil
+		case "json":
+			if isOther {
+				p := props[len(props)-1]
+				in = fmt.Sprintf("{%q: %q}", p, value)
+			}
+			err = json.Unmarshal([]byte(in), i)
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("unsupported format %s", format)
+		}
+	}
+	return errors.New("rpk bug, please describe how you encountered this at https://github.com/redpanda-data/redpanda/issues/new?assignees=&labels=kind%2Fbug&template=01_bug_report.md")
+}
+
+// getField deeply search in p for the value that reflect property props.
+func getField(props []string, p reflect.Value) (reflect.Value, reflect.Value, error) {
```

Comment:
I wonder if we should have a followup issue tracking a test that sets every possible field. There are a few cases that the below doesn't handle -- particularly, `*[]*type`. We shouldn't use a pointer to a slice of pointers ... but if we ever do, not great 😅